rustsec/src/cached_index.rs

```diff
@@ -36,47 +85,134 @@ impl CachedIndex {
+                Index::Git(rgi)
+            }
+            tame_index::index::ComboIndexCache::Sparse(si) => Index::SparseCached(si),
+        };
+
         Ok(CachedIndex {
             index,
             cache: Default::default(),
         })
     }
 
-    /// Load all version of the given crate from the crates.io index and put them into the cache
-    fn populate_cache(&mut self, package: &package::Name) -> Result<(), Error> {
-        let crate_releases = self.index.crate_(package.as_str()).ok_or_else(|| {
-            format_err!(
-                ErrorKind::NotFound,
-                "no such crate in the crates.io index: {}",
-                &package,
-            )
-        })?;
-
-        // We already loaded the full crate information, so populate all the versions in the cache
-        let versions: HashMap<String, bool> = crate_releases
-            .versions()
-            .iter()
-            .map(|v| (v.version().to_owned(), v.is_yanked()))
-            .collect();
-        self.cache.insert(package.to_owned(), versions);
+    /// Populates the cache entries for all of the specified crates
+    ///
+    /// This method is preferable to doing invidual updates via `cache_insert`/`is_yanked`
+    pub fn populate_cache(
+        &mut self,
+        packages: std::collections::BTreeSet<&package::Name>,
+    ) -> Result<(), Error> {
+        match &self.index {
+            Index::Git(_) | Index::SparseCached(_) => {
+                for pkg in packages {
+                    self.insert(pkg.to_owned(), self.index.krate(pkg));
+                }
+            }
+            Index::SparseRemote(rsi) => {
+                // Ensure we have a runtime
+                let rt = tame_index::external::tokio::runtime::Runtime::new().map_err(|err| {
+                    format_err!(
+                        ErrorKind::Registry,
+                        "unable to start a tokio runtime: {}",
+                        err
+                    )
+                })?;
+                let _rt = rt.enter();
+
+                /// This is the timeout per individual crate. If a crate fails to be
+                /// requested for a retriable reason then it will be retried until
+                /// this time limit is reached
+                const REQUEST_TIMEOUT: Option<std::time::Duration> =
+                    Some(std::time::Duration::from_secs(10));
+
+                let results = rsi
+                    .krates_blocking(
+                        packages
+                            .into_iter()
+                            .map(|p| p.as_str().to_owned())
+                            .collect(),
+                        true,
+                        REQUEST_TIMEOUT,
+                    )
+                    .map_err(|err| {
+                        format_err!(
+                            ErrorKind::Registry,
+                            "unable to acquire tokio runtime: {}",
```

Comment:
I believe this error message about Tokio runtime is misleading.

Reply:
This is handling an error returned by the function `krates_blocking`, right? For example, a timeout may cause it? Or am I misunderstanding the code?

Reply:
How so? In this case the error should never occur since the runtime is being created just above this function, but the error is only possible the in the case that https://docs.rs/tokio/1.29.1/tokio/runtime/struct.Handle.html#method.try_current fails.

Reply:
All timeouts or other errors would be particular to each individual crate with how the current code works.

Reply:
Ah! I see. I misunderstood the code then. Thank you for clarifying!